File: apps/mobile_companion/lib/protocol/word_filter.dart
import 'dart:convert';
import 'dart:typed_data';

import 'package:chirp_proto/proto/chat.pb.dart' as chat;

import 'hooks.dart';

/// 敏感词命中的两种客户端处置。没有 record:记录待审核是服务端职责
/// (服务端 WordFilter 已实现三级策略),客户端预检的职责只是不让脏词
/// 出门——要么改写,要么拦截。
enum WordFilterPolicy { replace, reject }

class WordFilterOptions {
  WordFilterOptions({
    this.terms = const <String>[],
    this.policy = WordFilterPolicy.replace,
    this.replacement = '**',
  });

  /// 词集合,大小写不敏感;构造时统一 lower + 去重。
  final List<String> terms;
  final WordFilterPolicy policy;
  final String replacement;
}

/// 敏感词预检拦截器:行为对齐服务端 chirp::chat::WordFilter 与 C++ core
/// (word_filter.h)/C#(WordFilterInterceptor.cs)的同名实现——词库格式、
/// ASCII 大小写不敏感子串匹配、mask 后重建的替换语义。只滤发送侧:接收
/// 内容信任服务端已按其策略处理。实例构造后不可变,可并发复用。
class WordFilterInterceptor extends MessageInterceptor {
  WordFilterInterceptor(WordFilterOptions options)
      : terms = parseWordLexicon(options.terms),
        _policy = options.policy,
        _replacement = options.replacement;

  /// lower-cased, deduplicated, sorted。
  final List<String> terms;
  final WordFilterPolicy _policy;
  final String _replacement;

  int get wordCount => terms.length;

  /// replace:命中时把改写后的内容写回 request.content 并放行;无命中
  /// 原样放行。reject:命中返回 false(管线抛 RequestError(blocked))。
  @override
  bool onBeforeSend(chat.SendMessageRequest request) {
    final content = utf8.decode(request.content);
    final lowered = asciiLower(content);
    final masked = _maskHits(lowered);
    if (masked == null) {
      return true;
    }
    if (_policy == WordFilterPolicy.reject) {
      return false;
    }
    request.content =
        Uint8List.fromList(utf8.encode(_rebuild(content, lowered, masked)));
    return true;
  }

  /// null = 无命中;否则为按 code unit 的命中区间 mask。
  List<bool>? _maskHits(String lowered) {
    if (terms.isEmpty) return null;
    final masked = List<bool>.filled(lowered.length, false);
    var hit = false;
    for (final term in terms) {
      // 词永不为空:parseWordLexicon 丢掉空行。
      var pos = lowered.indexOf(term);
      while (pos >= 0) {
        hit = true;
        for (var i = pos; i < pos + term.length; i++) {
          masked[i] = true;
        }
        pos = lowered.indexOf(term, pos + term.length);
      }
    }
    return hit ? masked : null;
  }

  /// 连续命中塌缩成一次替换,未命中字符保留原大小写。
  String _rebuild(String content, String lowered, List<bool> masked) {
    final filtered = StringBuffer();
    var i = 0;
    while (i < lowered.length) {
      if (masked[i]) {
        filtered.write(_replacement);
        while (i < lowered.length && masked[i]) {
          i++;
        }
      } else {
        filtered.writeCharCode(content.codeUnitAt(i));
        i++;
      }
    }
    return filtered.toString();
  }
}

/// 按服务端词库文件格式(每行一词,# 开头为注释,空行忽略)逐行解析。
/// trim 集合与服务端一致:空格与 CR/LF,不含 tab 等其余空白。
List<String> parseWordLexicon(List<String> lines) {
  final unique = <String>{};
  for (final raw in lines) {
    var term = raw;
    while (term.isNotEmpty &&
        (term.endsWith(' ') || term.endsWith('\r') || term.endsWith('\n'))) {
      term = term.substring(0, term.length - 1);
    }
    while (term.startsWith(' ')) {
      term = term.substring(1);
    }
    if (term.isEmpty || term.startsWith('#')) {
      continue;
    }
    unique.add(asciiLower(term));
  }
  final sorted = unique.toList()..sort();
  return sorted;
}

/// 只折叠 ASCII 大小写,与服务端 ToLower(默认 C locale)逐字节一致;
/// UTF-8 多字节序列不受影响,按 code unit(BMP 内与字节序一致)匹配。
String asciiLower(String text) {
  final out = StringBuffer();
  for (var i = 0; i < text.length; i++) {
    final c = text.codeUnitAt(i);
    out.writeCharCode(c >= 0x41 && c <= 0x5a ? c + 0x20 : c);
  }
  return out.toString();
}

File: apps/web_companion/src/protocol/word_filter.ts
/**
 * Sensitive-word pre-check interceptor — TypeScript face of the same
 * implementation the C++ core ships (sdks/core/include/chirp/word_filter.h)
 * and the Unity SDK (Runtime/Chirp/WordFilterInterceptor.cs), aligned with
 * the server-side chirp::chat::WordFilter: lexicon format, ASCII
 * case-insensitive substring matching, and the mask-then-rebuild replace
 * semantics. Send side only — receive content is trusted to the server's
 * own policy.
 */
import type { SendMessageRequest } from '@chirp/proto/chat';
import type { MessageInterceptor } from './hooks';

export type WordFilterPolicy = 'replace' | 'reject';

export interface WordFilterOptions {
  /** Terms, case-insensitive; lower-cased and de-duplicated on construction. */
  terms: readonly string[];
  /** Default 'replace'. There is no 'record': audit is the server's job. */
  policy?: WordFilterPolicy;
  /** Mask for replaced runs; default '**'. */
  replacement?: string;
}

/** Lexicon lines in the server file format (one term per line, '#' starts a
 * comment line, blank lines ignored, terms lower-cased and de-duplicated).
 * Feed the same lexicon file the server loads and both ends agree. */
export function parseWordLexicon(lines: readonly string[]): string[] {
  const unique = new Set<string>();
  for (const raw of lines) {
    // Trim exactly what the server trims: spaces and CR/LF — not the full
    // String.prototype.trim whitespace set.
    const term = raw.replace(/^[ ]+/, '').replace(/[ \r\n]+$/, '');
    if (term.length === 0 || term.startsWith('#')) {
      continue;
    }
    unique.add(asciiLower(term));
  }
  return [...unique].sort();
}

/** ASCII-only case folding, matching the server's ToLower (C locale,
 * byte-wise). UTF-8 multi-byte sequences pass through untouched. */
function asciiLower(text: string): string {
  let out = '';
  for (let i = 0; i < text.length; i++) {
    const code = text.charCodeAt(i);
    out += code >= 0x41 && code <= 0x5a ? String.fromCharCode(code + 0x20) : text[i];
  }
  return out;
}

export class WordFilterInterceptor implements MessageInterceptor {
  readonly wordCount: number;

  private readonly terms: string[]; // lower-cased, deduplicated, sorted
  private readonly policy: WordFilterPolicy;
  private readonly replacement: string;

  constructor(options: WordFilterOptions) {
    this.terms = parseWordLexicon(options.terms);
    this.policy = options.policy ?? 'replace';
    this.replacement = options.replacement ?? '**';
    this.wordCount = this.terms.length;
  }

  /** 'replace': hits are rewritten into req.content and the send proceeds;
   * clean text passes untouched. 'reject': a hit returns false — the
   * pipeline throws RequestError('blocked') and nothing reaches the wire. */
  onBeforeSend(request: SendMessageRequest): boolean {
    const content = new TextDecoder().decode(request.content);
    const lowered = asciiLower(content);
    const masked = this.maskHits(lowered);
    if (masked === null) {
      return true;
    }
    if (this.policy === 'reject') {
      return false;
    }
    request.content = new TextEncoder().encode(this.rebuild(content, lowered, masked));
    return true;
  }

  /** null = no hit; boolean[] = per-code-unit mask over the hit ranges. */
  private maskHits(lowered: string): boolean[] | null {
    if (this.terms.length === 0) {
      return null;
    }
    const masked = new Array<boolean>(lowered.length).fill(false);
    let hit = false;
    for (const term of this.terms) {
      // Terms are never empty: parseWordLexicon drops blank lines.
      let pos = lowered.indexOf(term);
      while (pos >= 0) {
        hit = true;
        masked.fill(true, pos, pos + term.length);
        pos = lowered.indexOf(term, pos + term.length);
      }
    }
    return hit ? masked : null;
  }

  /** Rebuild with adjacent hits collapsed into one replacement; characters
   * outside hit ranges keep their original casing. */
  private rebuild(content: string, lowered: string, masked: boolean[]): string {
    let filtered = '';
    for (let i = 0; i < lowered.length;) {
      if (masked[i]) {
        filtered += this.replacement;
        while (i < lowered.length && masked[i]) {
          i++;
        }
      } else {
        filtered += content[i];
        i++;
      }
    }
    return filtered;
  }
}
